Enable Core Data concurrency debug option in UI tests

crazytonyli opened the pull request:
The app crashes during UI tests when a Core Data concurrency issue occurs.

A Core Data concurrency issue is typically where a managed object is accessed outside of the context (or the queue managed by the context). Currently there are still many such issues in the app, post editing being the most impacted one, which is why the new
`crashOnCoreDataConcurrencyIssue` option is disabled in `EditorGutenbergTests`.

But, I think it's still valuable to ensure that those areas that don't have any Core Data concurrency issues to be maintained. And we should expose the new issue if one does occur. Doing that in UI tests is one easy way I can think of, which doesn't have a "blocking" factor—we can always disable the check for specific test(s).

## Regression Notes
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mokagio commented on WordPress/UITests/Utils/XCTest+Extensions.swift:
```diff
@@ -3,14 +3,22 @@ import XCTest
 
+        if crashOnCoreDataConcurrencyIssues {
+            app.launchArguments.append(contentsOf: ["-com.apple.CoreData.ConcurrencyDebug", "1"])
+        }
```
Only a subset of the tests use `setUpTestSuite`, which would result in this argument being applied inconsistently.

Have you considered using the test plan to set it and use this method to opt-out only?

<img width="918" alt="image" src="https://user-images.githubusercontent.com/1218433/227426670-e8c852d7-e2e2-465b-8134-f1b202275010.png">

crazytonyli replied:
Good suggestion. I hadn't thought about it, but I'll give it a go.

crazytonyli replied:
> Only a subset of the tests use `setUpTestSuite`

The idea is only turn the debug option on for _UI tests_, which all calls this funciton, I think?

mokagio replied:
> [UI tests] all calls this funciton, I think?

Oh, that is correct. I don't know why I got the impression only some of them called `setUpTestSuite` 🤔 But I checked each test and they all do.

Apologies.

mokagio reviewed (commented):
The UI tests on iPhone had 26 failures (waiting for #20378 to have better reporting on flakiness). I wonder how many of those were due to Core Data issues 🤔 

If there are many, I'd worry they would introduce too much noise in the report—Unless, that is, we committed to fix those issues ASAP.

crazytonyli commented:
> The UI tests on iPhone had 26 failures

Sorry, I should've mentioned in the PR description. All those failures should be addressed by https://github.com/wordpress-mobile/WordPress-iOS/pull/20394. I'll wait for that PR to be merged to re-run all the UI tests in this PR.

> [...] have better reporting on flakiness

I don't _think_ this change would cause flaky tests. I can later try re-run UI tests a few times to be sure.

mokagio commented:
@crazytonyli I see #20394 is not in this branch via [`66b4efc`](https://github.com/wordpress-mobile/WordPress-iOS/pull/20395/commits/66b4efc95e325fd0f59f62aa10809e26ee359e08). However, we still have more UI tests failures than other branches 🤔

crazytonyli commented:
I think the UI tests failed because they caught a known Core Data concurrency issue which should be fixed by https://github.com/wordpress-mobile/WordPress-iOS/pull/20413. Unfortunately, the CI job doesn't tell us the failure reason. The result bundle just says "timeout" because the app crashed. So, it would be good to attach a crash report to the result bundle if that's possible.

mokagio reviewed (approved):
Looks like all the tests failures we saw initially have been addressed by the PRs, as expected.

:shipit:

crazytonyli commented:
> I can later try re-run UI tests a few times to be sure.

Buildkite doesn't allow retrying a successful step. I think I'll have to amend the git history to re-build, which is what I'm going to do, for a few times, to be sure that this change won't cause UI test becoming flaky.

crazytonyli committed:
Enable Core Data concurrency debug option in UI tests

The app crashes during UI tests when a Core Data concurrency issue occurs.

A Core Data concurrency issue is typically where a managed object is
accessed outside of the context (or the queue managed by the context).
Currently there are still many such issues in the app, post editing
being the most impacted one, which is why the new
`crashOnCoreDataConcurrencyIssue` option is disabled in `EditorGutenbergTests`.

But, I think it's still valuable to ensure that those areas that don't
have any Core Data concurrency issues to be maintained. And we should
expose the new issue if one does occur. Doing that in UI tests is one
easy way I can think of, which doesn't have a "blocking" factor—we can
always disable the check for specific test(s).

mokagio commented:
> Buildkite doesn't allow retrying a successful step. I think I'll have to amend the git history to re-build, which is what I'm going to do, for a few times, to be sure that this change won't cause UI test becoming flaky.

I opened https://github.com/wordpress-mobile/WordPress-iOS/pull/20461 to run the UI tests step 6 times, which should increase our confidence. 

I'm also going to bump this milestone to 22.2 as I'm about to code freeze 22.1 and this is not a user facing PR so it doesn't matter whether it lands in 22.1 now or on `trunk` after the code freeze.

crazytonyli commented:
The CI jobs have been passing three times in a row, which I'll take as a confirmation that this PR doesn't cause flaky UI tests and merge this PR.

<img width="1145" alt="Screenshot 2023-04-03 at 1 00 37 PM" src="https://user-images.githubusercontent.com/1101828/229389694-dd272660-944f-4662-b9d2-0de512910afa.png">

mokagio commented:
@crazytonyli race condition 😄 

<img width="1090" alt="image" src="https://user-images.githubusercontent.com/1218433/229390024-3ac4d8ea-ef9a-4679-81c5-fc6f18b269d3.png">

Anyway, good choice. I should have checked to see if theres was something running in CI before going ahead and starting my experiment. I'll cancel it, then.

crazytonyli commented:
@mokagio Ops, sorry. That's a way smarter approach than my force-push 👍
